File: cpp/dsa/algorithms/sliding-window/permutation.cpp
bool checkInclusion(string target, string searchString) {
    if (target.length() > searchString.length()) return false;

    vector<int> targetFreq(26, 0);
    vector<int> windowFreq(26, 0);
    for (int i = 0; i < target.length(); i++) {
        targetFreq[target[i] - 'a']++;
    }

    int left = 0;
    for (int right = 0; right < searchString.length(); right++) {
        // 1️⃣ Expand
        windowFreq[searchString[right] - 'a']++;

        // 2️⃣ When window size reaches k
        if (right - left + 1 == target.length()) {
            // 3️⃣ Record / Evaluate
            if (checkMatch(targetFreq, windowFreq)) {
                return true;
            }

            // 4️⃣ Slide window
            windowFreq[searchString[left] - 'a']--;
            left++;
        }
    }
    return false;
}

bool checkMatch(const vector<int>& targetFreq, const vector<int>& windowFreq) {
    for (int j = 0; j < 26; j++) {
        if (targetFreq[j] != windowFreq[j]) return false;
    }
    return true;
}

File: python/dsa/algorithms/sliding-window/permutation.py
def check_inclusion(target: str, search_string: str) -> bool:
    if len(target) > len(search_string): return False

    target_freq = [0] * 26
    window_freq = [0] * 26
    for i in range(len(target)):
        target_freq[ord(target[i]) - ord('a')] += 1

    def check_match(c1, c2):
        for j in range(26):
            if c1[j] != c2[j]: return False
        return True

    left = 0
    for right in range(len(search_string)):
        # 1️⃣ Expand
        window_freq[ord(search_string[right]) - ord('a')] += 1

        # 2️⃣ When window size reaches k
        if right - left + 1 == len(target):
            # 3️⃣ Record / Evaluate
            if check_match(target_freq, window_freq):
                return True

            # 4️⃣ Slide window
            window_freq[ord(search_string[left]) - ord('a')] -= 1
            left += 1

    return False